packages/core/src/CardanoNode/util/bufferChainSyncEvent.ts

```diff
@@ -0,0 +1,50 @@
+import { Observable } from 'rxjs';
+import { RequestNext } from '../types';
+
+export interface WithRequestNext {
+  requestNext: RequestNext;
+}
+
+export const bufferChainSyncEvent =
+  <T extends WithRequestNext>(length: number) =>
+  (source$: Observable<WithRequestNext>) =>
+    new Observable<T>((subscriber) => {
+      const buffer: WithRequestNext[] = [];
+      let lastPull: RequestNext | undefined;
+      let unsubscribed = false;
+      let subscriberReady = true;
+
+      const next = () => {
+        if (unsubscribed) return;
+
+        if (buffer.length < length && lastPull) {
+          lastPull();
+          lastPull = undefined;
+        }
+
+        if (buffer.length === 0 || !subscriberReady) return;
+
+        const value = buffer.shift()!;
+
+        subscriberReady = false;
+        subscriber.next({
+          ...value,
+          requestNext: () => {
+            subscriberReady = true;
+            next();
+          }
+        } as T);
+      };
+
+      source$.subscribe({
+        complete: subscriber.complete.bind(subscriber),
+        error: subscriber.error.bind(subscriber),
+        next: (value) => {
+          lastPull = value.requestNext;
+          buffer.push(value);
+          next();
```

Comment:
This looks a little odd: are we triggering the emission when source$ emits? It should only do that when buffer is empty.